drivers/iio/light/cm36283.c

```diff
@@ -0,0 +1,852 @@
+					  enum iio_event_direction dir)
+{
+	struct cm36283_data *chip = iio_priv(indio_dev);
+
+	if (chan->type != IIO_PROXIMITY)
+		return -EINVAL;
+
+	switch (dir) {
+	case IIO_EV_DIR_FALLING:
+		return FIELD_GET(CM36283_PS_INT_OUT, chip->ps_conf[CM36283_PS_CONF1]);
+	case IIO_EV_DIR_RISING:
+		return FIELD_GET(CM36283_PS_INT_IN, chip->ps_conf[CM36283_PS_CONF1]);
+	default:
+		return -EINVAL;
+	}
+}
+
+static int cm36283_write_prox_event_config(struct iio_dev *indio_dev,
+					   const struct iio_chan_spec *chan,
+					   enum iio_event_type type,
+					   enum iio_event_direction dir,
+					   bool state)
+{
+	struct cm36283_data *chip = iio_priv(indio_dev);
+	struct i2c_client *client = chip->client;
+	int ret = 0;
+
+	if (chan->type != IIO_PROXIMITY)
+		return -EINVAL;
+
+	int new_ps_conf;
+
+	switch (dir) {
+	case IIO_EV_DIR_FALLING:
+		new_ps_conf = chip->ps_conf[CM36283_PS_CONF1] &
+			      ~CM36283_PS_INT_OUT;
+		new_ps_conf |= FIELD_PREP(CM36283_PS_INT_OUT, state);
+		break;
+	case IIO_EV_DIR_RISING:
+		new_ps_conf = chip->ps_conf[CM36283_PS_CONF1] &
+			      ~CM36283_PS_INT_IN;
+		new_ps_conf |= FIELD_PREP(CM36283_PS_INT_IN, state);
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	mutex_lock(&chip->lock);
+
+	ret = i2c_smbus_write_word_data(chip->client, CM36283_REG_PS_CONF1, new_ps_conf);
+	if (ret < 0)
+		dev_err(&client->dev,
+			"Failed to set proximity event interrupt config: %d\n", ret);
+	else
+		chip->ps_conf[CM36283_PS_CONF1] = new_ps_conf;
+
+	mutex_unlock(&chip->lock);
+
+	return ret;
+}
+
+static int cm36283_fallback_read_ps(struct iio_dev *indio_dev)
+{
+	struct cm36283_data *chip = iio_priv(indio_dev);
+	struct i2c_client *client = chip->client;
+	int data = i2c_smbus_read_word_data(client, CM36283_REG_PS_DATA);
+
+	if (data < 0)
+		return data;
+
+	if (data < chip->ps_away)
+		return IIO_EV_DIR_FALLING;
+	else if (data > chip->ps_close)
+		return IIO_EV_DIR_RISING;
+	else
+		return IIO_EV_DIR_NONE;
+}
+
+// Reading the interrupt flag acknowledges the interrupt
+static irqreturn_t cm36283_irq_handler(int irq, void *data)
+{
+	struct iio_dev *indio_dev = data;
+	struct cm36283_data *chip = iio_priv(indio_dev);
+	struct i2c_client *client = chip->client;
+	int ev_dir, ret;
+	u64 ev_code;
+
+	ret = i2c_smbus_read_word_data(client, CM36283_REG_INT_FLAG);
+	if (ret < 0) {
+		dev_err(&client->dev,
+			"Interrupt flag register read failed: %d\n", ret);
+		return IRQ_HANDLED;
+	}
+
+	ret >>= CM36283_INT_FLAG_SHIFT;
+	switch (ret) {
+	case CM36283_CLOSE:
+		ev_dir = IIO_EV_DIR_RISING;
+		break;
+	case CM36283_AWAY:
+		ev_dir = IIO_EV_DIR_FALLING;
+		break;
+	case CM36283_BOTH:
+		dev_warn(&client->dev,
+			"Interrupt flag was not cleared, reading proximity sensor data...\n");
+		ev_dir = cm36283_fallback_read_ps(indio_dev);
+		if (ev_dir < 0) {
+			dev_err(&client->dev, "Failed to settle interrupt state: %d\n", ret);
+			return IRQ_HANDLED;
+		}
+		break;
+	default:
+		dev_err(&client->dev, "Unknown interrupt state: %d\n", ret);
```

Comment:
```suggestion
		dev_err(&client->dev, "Unknown interrupt state: %x\n", ret);
```

nit: since this is a flag field, I think hex is easier to read